pywhy_stats/kci.py

```diff
@@ -0,0 +1,393 @@
+from functools import partial
+from typing import Callable, Optional, Tuple
+
+import numpy as np
+from numpy._typing import ArrayLike
+from scipy import stats
+from sklearn.metrics.pairwise import rbf_kernel
+
+from pywhy_stats.kernels import delta_kernel, estimate_squared_sigma_rbf
+from pywhy_stats.pvalue_result import PValueResult
+
+
+def ind(
+    X: np.ndarray,
+    Y: np.ndarray,
+    kernel_X: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+    kernel_Y: Optional[Callable[[np.ndarray], np.ndarray]] = None,
+    approx: bool = True,
+    null_sample_size: int = 1000,
+    threshold: float = 1e-5,
+    normalize_data: bool = True,
+) -> PValueResult:
+    """
+    Test whether X and Y are unconditionally independent using the Kernel Independence Tests.
+
+    For testing independence on continuous data, we leverage kernels :footcite:`Zhang2011` that
+    are computationally efficient.
+
+    Parameters
+    ----------
+    X : ArrayLike of shape (n_samples, n_features_x)
+        Data for variable X, which can be multidimensional.
+    Y : ArrayLike of shape (n_samples, n_features_y)
+        Data for variable Y, which can be multidimensional.
+    kernel_X : Callable[[ArrayLike], ArrayLike]
+        The kernel function for X. By default, the RBF kernel is used for continuous and the delta
+        kernel for categorical data.
```

Comment:
The issues w/ this is that categorical data does not necessarily have to be of type `str`. One can have for example: `X \in \{0, 1, 2, 3\}`, where X was transformed by sklearn's https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html. 

I'm inclined to just have RBF as the default and remove us guessing at categorical, but if you want to leave it, I would specifically state that categorical data in this function means `np.str` only.

Reply:
Ok, I will make that specific. Otherwise, I would say it is up to the user to use this correctly. When they perform any kind of transform (e.g. to discrete values without orders), then we cannot really catch that. In case of one-hot-encoders, that is even fine. The RBF kernel can be used with one-hot encoded features, it is just problematic if the mapped to numbers from 0 to n.